packages/process/process-monitor.pony

```diff
@@ -0,0 +1,382 @@
+use "files"
+
+use @pony_os_errno[I32]()
+use @pony_asio_event_create[AsioEventID](owner: AsioEventNotify, fd: U32, 
+      flags: U32, nsec: U64, noisy: Bool)
+use @pony_asio_event_unsubscribe[None](event: AsioEventID)
+use @pony_asio_event_destroy[None](event: AsioEventID)
+      
+primitive _EINTR        fun apply(): I32 => 4
+primitive _EAGAIN       fun apply(): I32 => 35
+primitive _STDINFILENO  fun apply(): U32 => 0
+primitive _STDOUTFILENO fun apply(): U32 => 1
+primitive _STDERRFILENO fun apply(): U32 => 2
+primitive _FSETFL       fun apply(): I32 => 4
+primitive _ONONBLOCK    fun apply(): I32 => 4
+
+primitive ExecveError
+primitive PipeError
+primitive Dup2Error
+primitive ForkError
+primitive FcntlError
+primitive WaitpidError
+primitive CloseError
+primitive ReadError
+primitive WriteError
+primitive Unsupported // we throw this on non POSIX systems
+
+type ProcessError is
+  ( ExecveError
+  | PipeError    
+  | Dup2Error    
+  | ForkError
+  | FcntlError
+  | WaitpidError
+  | CloseError
+  | ReadError
+  | WriteError
+  | Unsupported
+  )    
+
+  
+actor ProcessMonitor
+  """
+  Forks and monitors a process. Notifies a client about STDOUT / STDERR events.
+  """        
+  let _notifier: ProcessNotify
+  var _child_pid: I32 = -1
+  
+  var _stdout_event: AsioEventID = AsioEvent.none()
+  var _stderr_event: AsioEventID = AsioEvent.none()
+
+  var _read_buf: Array[U8] iso = recover Array[U8].undefined(4096) end
+
+  var _stdin_read:   U32 = -1
+  var _stdin_write:  U32 = -1
+  var _stdout_read:  U32 = -1
+  var _stdout_write: U32 = -1
+  var _stderr_read:  U32 = -1
+  var _stderr_write: U32 = -1
+    
+  var _stdout_open: Bool = true
+  var _stderr_open: Bool = true
+  
+  new create(notifier: ProcessNotify iso, path: String, 
+    args: Array[String] val, vars: Array[String] val) =>
+    """
+    Create infrastructure to communicate with a forked child process
+    and register the asio events. Fork child process and notify our
+    user about incoming data via the notifier. 
+    """
+    _notifier = consume notifier
+
+    ifdef posix then
+      try
+        (_stdin_read, _stdin_write)   = _make_pipe()
+        (_stdout_read, _stdout_write) = _make_pipe()
+        (_stderr_read, _stderr_write) = _make_pipe()
+      else
+        @close[I32](_stdin_read)
+        @close[I32](_stdin_write)
+        @close[I32](_stdout_read)
+        @close[I32](_stdout_write)
+        @close[I32](_stderr_read)
+        @close[I32](_stderr_write)
+        _notifier.failed(PipeError)
+      end
+      // fork child process
+      _child_pid = @fork[I32]()
+      match _child_pid
+      | -1  => _notifier.failed(ForkError)
+      | 0   => _child(path, args, vars)
+      else
+        _parent()
+      end
+    elseif windows then
+      _notifier.failed(Unsupported)
+    else
+      compile_error "unsupported platform"
+    end
+
+    
```

Comment:
Style-wise, there should probably be just one empty line between methods declared in the same type.

Also, I notice there are many empty lines at the end of this file - there should be only one (which GitHub will trim in the display because it's conventional to have on every file).